src/main/resources/assets/opencomputers/loot/openos/lib/core/full_keyboard.lua

```diff
@@ -1,133 +1,140 @@
 local keyboard = require("keyboard")
 
-keyboard.keys["1"]           = 0x02
```

Comment:
Regarding plan9k, we are planning to drop bundling it in OpenComputers 1.8.0 regardless. Don't worry too much about it, but it's good to use as a testcase for OC1 API compatibility.

From personal experience developing other things - the divide between key and character events is done because key scancodes (attached to physical key location) != character scancodes (affected by keymap, IME, etc).

I'm not sure how LWJGL 3's specific implementation works, but this might be a good first try:

* For key events, ignore all scancodes which map to *printable* characters (ASCII range 32-126 is a good first start).
* For character events, ignore all events which map to *non-printable* characters (ASCII range 0-31).
* Create mapping tables from key scancodes to characters where applicable, and from characters to key scancodes where applicable. (See https://github.com/MightyPirates/OpenComputers/blob/master-MC1.7.10/src/main/resources/assets/opencomputers/loot/openos/lib/core/full_keyboard.lua for more information). 
* Tentatively ignore character events with characters outside the 32-126 range. Add a TODO comment to examine interactions with IMEs (for, say, typing Japanese directly into the OpenComputers machine). This is best skipped for a first alpha.

Reply:
> I'm not sure how LWJGL 3's specific implementation works, but this might be a good first try:

LGJGL 3 just uses [GLFW key codes](https://www.glfw.org/docs/latest/group__keys.html), so it should be possible to make an exhaustive list and map between the two.

Afraid I didn't bother to do this for ComputerCraft or I'd share mine. Sorry!

Reply:
Hmm, I'm reading the LWJGL 2 code currently. At least on windows, it seems to be unable to report a non-documented keycode. In addition, it appears to be doing the exact "correlate key and character events" wizardry I was afraid of implementing myself. I'll have to look into this more, and whether it works for other OSes too (and how to avoid stuck keys).

Regarding mapping scancodes to characters and back, I don't think that's feasible. I'd have to know the keyboard layout (every keyboard layout) to be able to do this correctly, just like a physical OS would. And I don't have the keyboards to test it.

Reply:
Tricky.

I do agree that separating key and text input events OC-side going forward would be a good idea; possibly by introducing two new events and deprecating the old one, then emulating the new events in 1.7 and 1.12 as well. This is because once you add IMEs to the mix, there is no longer a correlation between text input and keypresses (LWJGL2 just didn't support IMEs in the first place).

However, what would be the best way to do that while minimizing impact to legacy code? My proposal: We keep treating key_down/key_up as scancode events, with a scancode-derived QWERTY-assuming character value for "char". We also add a text_input event. On 1.7/1.12, key_down also implies emitting a matching text_input event.

Reply:
I talked to payonel, and he seems to agree that a new `text_input` event + a non-ideal compatibility layer for `key_down` would be the best idea. See [this](https://paste.asie.pl/KPtr) for a simple patch to emulate `text_input` on < 1.13. Do you think that would be sufficient?

Reply:
I'm in favor of a `text_input` and GLFW conveniently uses ASCII codes for printable keys (32-126) which I can enhance with a crude shift-to-uppercase check. It'll still be wrong on non-US keyboards though. Would it be OK if I continue firing 2 `key_down` events for characters, one with only the key and one with only the character (in addition to the new `text_input`)?

Reply:
My opinion is that `key_down` should *only* react to scancode events (and translate them to characters for some subset of keys), while `text_input` should *only* react to character events. The idea is that code which uses only `key_down` will work well enough for the time being, and eventually software can be migrated to use `text_input` for printables.

Reply:
Now, if you really want to go the extra mile, you could make character events in the 32-126 range emulate `key_down` calls on top of emitting `text_input`, while scancode events which result in a 32-126 range character are ignored. This will solve most of the keymap problem.

Reply:
I've looked into manual translation to/from characters again and I'm even more convinced that it's not possible without a key map (GLFW doesn't provide one) as it becomes impossible to type symbols which requires a modifier key, even on US keyboards. Instead I've rewritten the system to wait for character events on all keys, joining the first character with the most recent key (following characters are dropped, as in 1.12.2). Finally, during a GUI tick the client sends key events that weren't associated with any character (doing this after events are polled would be nice, but there's no Forge hook for it).
Additionally, Minecraft's GUIs can only take a char (not a code point) but `KeyboardListener` splits them up correctly, so I've taken the liberty of rejoining high & low surrogate pairs similar to keys & characters (for `text_input` only, surrogates don't get delivered for `key_down` and `key_up` at all). The reason for doing this is that in UTF-8, non-BMP is encoded differently from the concatenation of a high and low surrogate - the latter isn't even valid UTF-8. Note that I do not have an IME to test this with non-BMP codes.

Reply:
This is all very good. I wasn't worried about symbols which require a modifier key, as I'd expect common applications (like `edit` and the shell) to be patched to properly support `text_input` - however, I'm impressed that you went the full mile to support it. Note that until you pull in the latest `MC1.12` changes, OpenComputers will *not* correctly parse non-BMP key codes in all cases.

Reply:
I've briefly reviewed the PR, and I like it - my only concern is that whether or not it is possible for the key and char diverge significantly (say, is it possible for me to press 'A' and 'B' simultaneously in such a way that the char for 'A' and the scancode for 'B' gets emitted, or vice versa?), and how rare such an occurence would be.

Reply:
I don't think that's an issue, key and char callbacks are only called while the main thread runs `GLFW.glfwPollEvents` and are executed in whichever order the OS delivers them. Specifically (on windows), when you [translate](https://learn.microsoft.com/en-us/windows/win32/inputdev/using-keyboard-input#translating-character-messages) a key code into a character, that event is always delivered next so unless GLFW mixes them it's safe to correlate them.
What does happen (even in LWJGL 2 / OC 1.12.2) is that pressing an accent key (which doesn't generate a character immediately) followed by a non-modifier key will result in the accent replacing whatever character you would have gotten. This results in surreal combinations where the enter `key_down` event has the right key code, but the accent character. Same method, same... feature I guess.

Reply:
Okay. In that case, I think we're good.